File: src/main/java/hardcoded/aoc2021/Day9.java
package hardcoded.aoc2021;

import java.util.*;

import hardcoded.util.Utils;

public class Day9 {
	public static void main(String[] args) throws Exception {
		List<int[]> lines = Utils.readAllLines(2021, "day9").stream().map(s -> s.chars().map(i -> (int)(i - '0')).toArray()).toList();
		int width = lines.get(0).length;
		int height = lines.size();
		
		Utils.printf("Day 9\n");
		Utils.printf("PartOne: %d\n", partOne(lines, width, height));
		Utils.printf("PartTwo: %d\n", partTwo(lines, width, height));
	}
	
	public static int getLowPoint(List<int[]> lines, int x, int y, int width, int height) {
		int value = lines.get(y)[x];
		
		return ((x > 0 && value >= lines.get(y)[x - 1])
		|| (x < width - 1 && value >= lines.get(y)[x + 1])
		|| (y > 0 && value >= lines.get(y - 1)[x])
		|| (y < height - 1 && value >= lines.get(y + 1)[x])) ? 0:(value + 1);
	}

	private static long getCoordinate(int x, int y) {
		return ((long)(x) & 0xffffffffL) | (((long)y) << 32L);
	}
	
	private static int getBasinSize(List<int[]> lines, Set<Long> set, int x, int y, int width, int height) {
		int value = lines.get(y)[x];
		// A basin of value 9 does not count.
		if(value == 9) return 0;
		
		// This position has already been computed.
		if(!set.add(getCoordinate(x, y))) return 0;
		
		int count = 1;
		if(x > 0 && value < lines.get(y)[x - 1]) {
			count += getBasinSize(lines, set, x - 1, y, width, height);
		}
		
		if(x < width - 1 && value < lines.get(y)[x + 1]) {
			count += getBasinSize(lines, set, x + 1, y, width, height);
		}
		
		if(y > 0 && value < lines.get(y - 1)[x]) {
			count += getBasinSize(lines, set, x, y - 1, width, height);
		}
		
		if(y < height - 1 && value < lines.get(y + 1)[x]) {
			count += getBasinSize(lines, set, x, y + 1, width, height);
		}
		
		return count;
	}
	
	public static int partOne(List<int[]> lines, int width, int height) throws Exception {
		int score = 0;
		for(int y = 0; y < lines.size(); y++) {
			for(int x = 0; x < width; x++) {
				score += getLowPoint(lines, x, y, width, height);
			}
		}
		
		return score;
	}
	
	public static int partTwo(List<int[]> lines, int width, int height) throws Exception {
		List<Integer> basinSizes = new ArrayList<>();
		Set<Long> set = new HashSet<>();
		
		for(int y = 0; y < lines.size(); y++) {
			for(int x = 0; x < width; x++) {
				// If we found a low point we found a basin.
				int lowPoint = getLowPoint(lines, x, y, width, height);
				if(lowPoint > 0) {
					// Find all numbers next to this that is one above
					basinSizes.add(getBasinSize(lines, set, x, y, width, height));
					set.clear();
				}
			}
		}
		
		basinSizes.sort(Collections.reverseOrder());
		return basinSizes.get(0) * basinSizes.get(1) * basinSizes.get(2);
	}
}

File: src/main/python/Day9.py
import Utils as Utils

def getLowPoint(lines, x, y, width, height):
    value = lines[y][x]
    
    return 0 if ((x > 0 and value >= lines[y][x - 1])
    or (x < width - 1 and value >= lines[y][x + 1])
    or (y > 0 and value >= lines[y - 1][x])
    or (y < height - 1 and value >= lines[y + 1][x])) else value + 1

def getBasinSize(lines, coordinateSet, x, y, width, height):
    value = lines[y][x]
    if value == 9:
        return 0
    
    if (x, y) in coordinateSet:
        return 0
    coordinateSet.add((x, y))
    
    count = 1
    if x > 0 and value < lines[y][x - 1]:
        count += getBasinSize(lines, coordinateSet, x - 1, y, width, height)
    if x < width - 1 and value < lines[y][x + 1]:
        count += getBasinSize(lines, coordinateSet, x + 1, y, width, height)
    if y > 0 and value < lines[y - 1][x]:
        count += getBasinSize(lines, coordinateSet, x, y - 1, width, height)
    if y < height - 1 and value < lines[y + 1][x]:
        count += getBasinSize(lines, coordinateSet, x, y + 1, width, height)
    
    return count

@Utils.time_it_and_evaluate
def partOne(lines, width, height):
    score = 0
    for y in range(height):
        for x in range(width):
            score += getLowPoint(lines, x, y, width, height)
    return score

@Utils.time_it_and_evaluate
def partTwo(lines, width, height):
    basinSizes = []
    coordinateSet = set()
    
    for y in range(height):
        for x in range(width):
            lowPoint = getLowPoint(lines, x, y, width, height)
            if lowPoint > 0:
                basinSizes.append(getBasinSize(lines, coordinateSet, x, y, width, height))
                coordinateSet.clear()
    basinSizes.sort(reverse=True)
    return basinSizes[0] * basinSizes[1] * basinSizes[2]

if __name__ == '__main__':
    lines = [ [ int(i) for i in x ] for x in Utils.readAllLines(2021, 'day9') ]
    width = len(lines[0])
    height = len(lines)
    
    print('Day 9')
    partOne(lines, width, height)
    partTwo(lines, width, height)